[Feature]: Add directory controls

### Feature request

Add capabilities to create new directories, new files, amend the name of existing directories and files, and delete directories and files from within the application. Possibly also the capability to hide (but not delete) files that aren't relevant to an application (making use of an ignore list in `tnt.config.json`).

This was originally planned for #83.

So that's...

- [ ] Delete files/subdirectories
- [ ] Amend file/subdirectory names
- [ ] Create subdirectories
- [ ] Create files
- [ ] Hide irrelevant subdirectories/files

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

I've loosely tried following this: https://github.com/nuxt/nuxt/discussions/26942

...but the result now is... `Error: Cannot find module '@stroncium/procfs'`

This may be a no-go.

Some general things I do know:

1. The developer of trash is predominantly a Mac dev; that's probably where this is best tested
2. This issue is Linux, but I don't just want to omit the feature from Linux if we can help that
3. I don't necessarily know that it works on Windows - I still need to setup an appropriate means of testing Windows dev builds locally

I don't want to keep banging my head against this to achieve... a potential future feature (bulk trash operations). We can get the feature we're currently aiming for (single file deletion) another way - using Electron shell. So let's abandon what I'm currently doing and reroute down that path instead.

--- a/packages/core/composables/tntApi.ts
+++ b/packages/core/composables/tntApi.ts
@@ -58,6 +58,12 @@ export const useTntApi = () => {
         return await window.api.send('update-file', file, content)
       } // TODO: else - should we attempt localStorage/IndexedDB/Dexie here?
       return
+    },
+    createFolder: async (folder) => {
+      if (isElectron()) {
+        return await window.api.send('create-folder', folder)
+      } // TODO: else - should we attempt localStorage/IndexedDB/Dexie here?
+      return
     }
   }
 }

--- a/packages/core/composables/tntApi.ts
+++ b/packages/core/composables/tntApi.ts
@@ -35,6 +35,12 @@ export const useTntApi = () => {
       }
       return
     },
+    renameFile: async (file:string, name:string) => {
+      if (isElectron()) {
+        return await window.api.invoke('rename-file', file, name)
+      }
+      return
+    },
     deleteFile: async (file:string) => {
       if (isElectron()) {
         return await window.api.invoke('delete-file', file)

--- a/packages/core/composables/tntApi.ts
+++ b/packages/core/composables/tntApi.ts
@@ -53,7 +53,7 @@ export const useTntApi = () => {
       } // TODO: else - should we attempt localStorage/IndexedDB/Dexie here?
       return
     },
-    updateFile: async (file, content) => {
+    updateFile: async (file, content?) => {
       if (isElectron()) {
         return await window.api.send('update-file', file, content)
       } // TODO: else - should we attempt localStorage/IndexedDB/Dexie here?

--- a/packages/core/composables/tntApi.ts
+++ b/packages/core/composables/tntApi.ts
@@ -35,6 +35,12 @@ export const useTntApi = () => {
       }
       return
     },
+    deleteFile: async (file:string) => {
+      if (isElectron()) {
+        return await window.api.invoke('delete-file', file)
+      }
+      return
+    },
     loadFile: async (file) => {
       if (isElectron()) {
         return await window.api.invoke('load-file', file)

--- a/packages/electron/components/global/TntElectron/Directory.vue
+++ b/packages/electron/components/global/TntElectron/Directory.vue
@@ -1,4 +1,7 @@
 <script setup>
+import { useDirectoryStore } from '@/stores/directory'
+import { storeToRefs } from 'pinia'
+
 const props = defineProps({
   path: {
     type: String,
@@ -16,17 +19,18 @@ const props = defineProps({
   }
 })
 
-const files = ref([])
+// Store
+const store = useDirectoryStore()
+// Store: State/Getters
+const { tree } = storeToRefs(store)
+// Store: Actions
+const { fetchDirectory } = store
 
-async function fetchFiles() {
-  const dir = await useTntApi().listFiles(props.path, { exclude: props.exclude, extensions: props.extensions, depth: props.depth })
-  if (dir) files.value = dir
-}
-
-fetchFiles()
+// Init
+fetchDirectory(props.path, { exclude: props.exclude, extensions: props.extensions, depth: props.depth })
 </script>
 
 <template lang="pug">
-slot(:files="files")
-  TntElectronDirectoryTree(:rootDir="files.path" :files="files" :open="true" :filter="filter" :links="links")
+slot(:files="tree")
+  TntElectronDirectoryTree(:rootDir="tree.path" :files="tree" :open="true" :filter="filter" :links="links")
 </template>

--- a/packages/electron/components/global/TntElectron/Directory.vue
+++ b/packages/electron/components/global/TntElectron/Directory.vue
@@ -32,5 +32,5 @@ fetchDirectory(props.path, { exclude: props.exclude, extensions: props.extension
 
 <template lang="pug">
 slot(:files="tree")
-  TntElectronDirectoryTree(:rootDir="tree.path" :files="tree" :open="true" :filter="filter" :links="links")
+  TntElectronDirectoryTree(:files="tree" :open="true" :filter="filter" :links="links")
 </template>

--- a/packages/electron/components/global/TntElectron/DirectoryTree.vue
+++ b/packages/electron/components/global/TntElectron/DirectoryTree.vue
@@ -17,22 +17,42 @@ const store = useDirectoryStore()
 // Getters
 const { root } = storeToRefs(store)
 // Store: Actions
-const { createFile, deleteFile } = store
+const { createFile, createFolder, deleteFile } = store
 
 const children = computed(() => {
   let filtered = props.filter ? props.files.children?.filter(f => f.children || props.filter.test(f.name)) : props.files.children
   return _sortBy(filtered, [function(o) { return !o.children }, 'name'])
 })
 
+// Adding new files
 const newFile = ref('')
+const addingNewFile = ref(false)
 
-function create() {
+function createNewFile() {
   createFile(`${props.files.path.replace(`${root.value}/`, '')}/${newFile.value}`)
   newFile.value = ""
+  addingNewFile.value = false
 }
 
-function cancel() {
+function cancelNewFile() {
   newFile.value = ""
+  addingNewFile.value = false
+}
+
+// Adding new folders
+const newFolder = ref('')
+const addingNewFolder = ref(false)
+
+function createNewFolder() {
+  // TODO: Should this be a separate store action?
+  createFolder(`${props.files.path.replace(`${root.value}/`, '')}/${newFolder.value}`)
+  newFolder.value = ""
+  addingNewFolder.value = false
+}
+
+function cancelNewFolder() {
+  newFolder.value = ""
+  addingNewFolder.value = false
 }
 </script>
 
@@ -57,8 +77,19 @@ details(:open="open")
           Icon(name="fa6-solid:trash-can")
       TntElectronDirectoryTreeFileName(v-else v-model:path="child.path" v-model:name="child.name")
     li
-      TntForm.inline(@submit="create()" @keydown.esc="cancel()")
+      TntButton.btn-none(v-if="!addingNewFile" @click="addingNewFile = true")
+        span New file
+        Icon.ml-2(name="fa6-solid:file")
+      TntForm.inline(v-else @submit="createNewFile()" @keydown.esc="cancelNewFile()")
         TntInput.inline-block(v-model="newFile")
+        TntSubmit.btn-none.ml-2
+          Icon(name="fa6-solid:plus")
+    li
+      TntButton.btn-none(v-if="!addingNewFolder" @click="addingNewFolder = true")
+        span New folder
+        Icon.ml-2(name="fa6-solid:folder")
+      TntForm.inline(v-else @submit="createNewFolder()" @keydown.esc="cancelNewFolder()")
+        TntInput.inline-block(v-model="newFolder")
         TntSubmit.btn-none.ml-2
           Icon(name="fa6-solid:plus")
 </template>

--- a/packages/electron/components/global/TntElectron/DirectoryTree.vue
+++ b/packages/electron/components/global/TntElectron/DirectoryTree.vue
@@ -1,5 +1,6 @@
 <script setup>
 import { useDirectoryStore } from '@/stores/directory'
+import { storeToRefs } from 'pinia'
 
 const props = defineProps({
   rootDir: String,
@@ -14,6 +15,8 @@ const props = defineProps({
 
 // Store
 const store = useDirectoryStore()
+// Getters
+const { root } = storeToRefs(store)
 // Store: Actions
 const { deleteFile } = store
 
@@ -27,8 +30,8 @@ const children = computed(() => {
 details(:open="open")
   summary.cursor-pointer
     span {{ files.name }}
-    TntButton.btn-none.float-right(@click="deleteFile(files.path.replace(`${rootDir}/`, ''))" class="text-danger-light hover:text-danger-light-hover dark:text-danger-dark dark:hover:text-danger-dark-hover")
-        Icon(name="fa6-solid:trash-can")
+    TntButton.btn-none.float-right(v-if="files.path !== root" @click="deleteFile(files.path.replace(`${rootDir}/`, ''))" class="text-danger-light hover:text-danger-light-hover dark:text-danger-dark dark:hover:text-danger-dark-hover")
+      Icon(name="fa6-solid:trash-can")
   ul.ml-4
     li(v-for="child in children" :key="child.path")
       component(

--- a/packages/electron/components/global/TntElectron/DirectoryTree.vue
+++ b/packages/electron/components/global/TntElectron/DirectoryTree.vue
@@ -29,7 +29,8 @@ const children = computed(() => {
 <template lang="pug">
 details(:open="open")
   summary.cursor-pointer
-    span {{ files.name }}
+    TntElectronDirectoryTreeFileName(v-if="files.path !== root" :path="files.path" :name="files.name")
+    span(v-else) {{ files.name }}
     TntButton.btn-none.float-right(v-if="files.path !== root" @click="deleteFile(files.path.replace(`${rootDir}/`, ''))" class="text-danger-light hover:text-danger-light-hover dark:text-danger-dark dark:hover:text-danger-dark-hover")
       Icon(name="fa6-solid:trash-can")
   ul.ml-4
@@ -41,10 +42,11 @@ details(:open="open")
         :files="child"
       )
       span(v-else-if="links")
-        NuxtLink(:to="child.path.replace(`${rootDir}/`, '')")  {{ child.name }}
+        NuxtLink(:to="child.path.replace(`${rootDir}/`, '')")
+          TntElectronDirectoryTreeFileName(:path="child.path" :name="child.name")
         TntButton.btn-none.float-right(@click="deleteFile(child.path.replace(`${rootDir}/`, ''))" class="text-danger-light hover:text-danger-light-hover dark:text-danger-dark dark:hover:text-danger-dark-hover")
           Icon(name="fa6-solid:trash-can")
-      span(v-else)  {{ child.name }}
+      TntElectronDirectoryTreeFileName(v-else :path="child.path" :name="child.name")
 </template>
 
 <style lang="postcss" scoped>

--- a/packages/electron/components/global/TntElectron/DirectoryTree.vue
+++ b/packages/electron/components/global/TntElectron/DirectoryTree.vue
@@ -28,7 +28,7 @@ const children = computed(() => {
 <template lang="pug">
 details(:open="open")
   summary.cursor-pointer
-    TntElectronDirectoryTreeFileName(v-if="files.path !== root" :path="files.path" :name="files.name")
+    TntElectronDirectoryTreeFileName(v-if="files.path !== root" v-model:path="files.path" v-model:name="files.name")
     span(v-else) {{ files.name }}
     TntButton.btn-none.float-right(v-if="files.path !== root" @click="deleteFile(files.path.replace(`${root}/`, ''))" class="text-danger-light hover:text-danger-light-hover dark:text-danger-dark dark:hover:text-danger-dark-hover")
       Icon(name="fa6-solid:trash-can")
@@ -41,10 +41,10 @@ details(:open="open")
         :files="child"
       )
       span(v-else-if="links")
-        TntElectronDirectoryTreeFileName(:path="child.path" :name="child.name" :link="links")
+        TntElectronDirectoryTreeFileName(v-model:path="child.path" v-model:name="child.name" :link="links")
         TntButton.btn-none.float-right(@click="deleteFile(child.path.replace(`${root}/`, ''))" class="text-danger-light hover:text-danger-light-hover dark:text-danger-dark dark:hover:text-danger-dark-hover")
           Icon(name="fa6-solid:trash-can")
-      TntElectronDirectoryTreeFileName(v-else :path="child.path" :name="child.name")
+      TntElectronDirectoryTreeFileName(v-else v-model:path="child.path" v-model:name="child.name")
 </template>
 
 <style lang="postcss" scoped>

--- a/packages/electron/components/global/TntElectron/DirectoryTree.vue
+++ b/packages/electron/components/global/TntElectron/DirectoryTree.vue
@@ -3,7 +3,6 @@ import { useDirectoryStore } from '@/stores/directory'
 import { storeToRefs } from 'pinia'
 
 const props = defineProps({
-  rootDir: String,
   files: Object,
   open: {
     type: Boolean,
@@ -31,7 +30,7 @@ details(:open="open")
   summary.cursor-pointer
     TntElectronDirectoryTreeFileName(v-if="files.path !== root" :path="files.path" :name="files.name")
     span(v-else) {{ files.name }}
-    TntButton.btn-none.float-right(v-if="files.path !== root" @click="deleteFile(files.path.replace(`${rootDir}/`, ''))" class="text-danger-light hover:text-danger-light-hover dark:text-danger-dark dark:hover:text-danger-dark-hover")
+    TntButton.btn-none.float-right(v-if="files.path !== root" @click="deleteFile(files.path.replace(`${root}/`, ''))" class="text-danger-light hover:text-danger-light-hover dark:text-danger-dark dark:hover:text-danger-dark-hover")
       Icon(name="fa6-solid:trash-can")
   ul.ml-4
     li(v-for="child in children" :key="child.path")
@@ -42,9 +41,8 @@ details(:open="open")
         :files="child"
       )
       span(v-else-if="links")
-        NuxtLink(:to="child.path.replace(`${rootDir}/`, '')")
-          TntElectronDirectoryTreeFileName(:path="child.path" :name="child.name")
-        TntButton.btn-none.float-right(@click="deleteFile(child.path.replace(`${rootDir}/`, ''))" class="text-danger-light hover:text-danger-light-hover dark:text-danger-dark dark:hover:text-danger-dark-hover")
+        TntElectronDirectoryTreeFileName(:path="child.path" :name="child.name" :link="links")
+        TntButton.btn-none.float-right(@click="deleteFile(child.path.replace(`${root}/`, ''))" class="text-danger-light hover:text-danger-light-hover dark:text-danger-dark dark:hover:text-danger-dark-hover")
           Icon(name="fa6-solid:trash-can")
       TntElectronDirectoryTreeFileName(v-else :path="child.path" :name="child.name")
 </template>

--- a/packages/electron/components/global/TntElectron/DirectoryTree.vue
+++ b/packages/electron/components/global/TntElectron/DirectoryTree.vue
@@ -27,6 +27,9 @@ const children = computed(() => {
 // Adding new files
 const newFile = ref('')
 const addingNewFile = ref(false)
+const newFileRules = ref({
+  matches: props.filter
+})
 
 function createNewFile() {
   createFile(`${props.files.path.replace(`${root.value}/`, '')}/${newFile.value}`)
@@ -81,7 +84,7 @@ details(:open="open")
         span New file
         Icon.ml-2(name="fa6-solid:file")
       TntForm.inline(v-else @submit="createNewFile()" @keydown.esc="cancelNewFile()")
-        TntInput.inline-block(v-model="newFile")
+        TntInput.inline-block(v-model="newFile" :rules="newFileRules")
         TntSubmit.btn-none.ml-2
           Icon(name="fa6-solid:plus")
     li

--- a/packages/electron/components/global/TntElectron/DirectoryTree.vue
+++ b/packages/electron/components/global/TntElectron/DirectoryTree.vue
@@ -17,12 +17,23 @@ const store = useDirectoryStore()
 // Getters
 const { root } = storeToRefs(store)
 // Store: Actions
-const { deleteFile } = store
+const { createFile, deleteFile } = store
 
 const children = computed(() => {
   let filtered = props.filter ? props.files.children?.filter(f => f.children || props.filter.test(f.name)) : props.files.children
   return _sortBy(filtered, [function(o) { return !o.children }, 'name'])
 })
+
+const newFile = ref('')
+
+function create() {
+  createFile(`${props.files.path.replace(`${root.value}/`, '')}/${newFile.value}`)
+  newFile.value = ""
+}
+
+function cancel() {
+  newFile.value = ""
+}
 </script>
 
 <template lang="pug">
@@ -45,6 +56,11 @@ details(:open="open")
         TntButton.btn-none.float-right(@click="deleteFile(child.path.replace(`${root}/`, ''))" class="text-danger-light hover:text-danger-light-hover dark:text-danger-dark dark:hover:text-danger-dark-hover")
           Icon(name="fa6-solid:trash-can")
       TntElectronDirectoryTreeFileName(v-else v-model:path="child.path" v-model:name="child.name")
+    li
+      TntForm.inline(@submit="create()" @keydown.esc="cancel()")
+        TntInput.inline-block(v-model="newFile")
+        TntSubmit.btn-none.ml-2
+          Icon(name="fa6-solid:plus")
 </template>
 
 <style lang="postcss" scoped>

--- a/packages/electron/components/global/TntElectron/DirectoryTree.vue
+++ b/packages/electron/components/global/TntElectron/DirectoryTree.vue
@@ -46,3 +46,9 @@ details(:open="open")
           Icon(name="fa6-solid:trash-can")
       span(v-else)  {{ child.name }}
 </template>
+
+<style lang="postcss" scoped>
+details li {
+  @apply hover:bg-base-900/50;
+}
+</style>

--- a/packages/electron/components/global/TntElectron/DirectoryTree.vue
+++ b/packages/electron/components/global/TntElectron/DirectoryTree.vue
@@ -1,4 +1,6 @@
 <script setup>
+import { useDirectoryStore } from '@/stores/directory'
+
 const props = defineProps({
   rootDir: String,
   files: Object,
@@ -10,6 +12,11 @@ const props = defineProps({
   filter: RegExp,
 })
 
+// Store
+const store = useDirectoryStore()
+// Store: Actions
+const { deleteFile } = store
+
 const children = computed(() => {
   let filtered = props.filter ? props.files.children?.filter(f => f.children || props.filter.test(f.name)) : props.files.children
   return _sortBy(filtered, ['children', 'name'])
@@ -18,7 +25,10 @@ const children = computed(() => {
 
 <template lang="pug">
 details(:open="open")
-  summary.cursor-pointer {{ files.name }}
+  summary.cursor-pointer
+    span {{ files.name }}
+    TntButton.btn-none.float-right(@click="deleteFile(files.path.replace(`${rootDir}/`, ''))" class="text-danger-light hover:text-danger-light-hover dark:text-danger-dark dark:hover:text-danger-dark-hover")
+        Icon(name="fa6-solid:trash-can")
   ul.ml-4
     li(v-for="child in children")
       component(
@@ -27,6 +37,9 @@ details(:open="open")
         v-bind="_omit(props, ['files', 'open'])"
         :files="child"
       )
-      NuxtLink(v-else-if="links" :to="child.path.replace(`${rootDir}/`, '')" :key="child.path")  {{ child.name }}
+      span(v-else-if="links")
+        NuxtLink(:to="child.path.replace(`${rootDir}/`, '')" :key="child.path")  {{ child.name }}
+        TntButton.btn-none.float-right(@click="deleteFile(child.path.replace(`${rootDir}/`, ''))" class="text-danger-light hover:text-danger-light-hover dark:text-danger-dark dark:hover:text-danger-dark-hover")
+          Icon(name="fa6-solid:trash-can")
       span(v-else :key="child.path")  {{ child.name }}
 </template>

--- a/packages/electron/components/global/TntElectron/DirectoryTree.vue
+++ b/packages/electron/components/global/TntElectron/DirectoryTree.vue
@@ -30,7 +30,7 @@ details(:open="open")
     TntButton.btn-none.float-right(@click="deleteFile(files.path.replace(`${rootDir}/`, ''))" class="text-danger-light hover:text-danger-light-hover dark:text-danger-dark dark:hover:text-danger-dark-hover")
         Icon(name="fa6-solid:trash-can")
   ul.ml-4
-    li(v-for="child in children")
+    li(v-for="child in children" :key="child.path")
       component(
         v-if="child.children"
         is="tnt-electron-directory-tree"
@@ -38,8 +38,8 @@ details(:open="open")
         :files="child"
       )
       span(v-else-if="links")
-        NuxtLink(:to="child.path.replace(`${rootDir}/`, '')" :key="child.path")  {{ child.name }}
+        NuxtLink(:to="child.path.replace(`${rootDir}/`, '')")  {{ child.name }}
         TntButton.btn-none.float-right(@click="deleteFile(child.path.replace(`${rootDir}/`, ''))" class="text-danger-light hover:text-danger-light-hover dark:text-danger-dark dark:hover:text-danger-dark-hover")
           Icon(name="fa6-solid:trash-can")
-      span(v-else :key="child.path")  {{ child.name }}
+      span(v-else)  {{ child.name }}
 </template>

--- a/packages/electron/components/global/TntElectron/DirectoryTree.vue
+++ b/packages/electron/components/global/TntElectron/DirectoryTree.vue
@@ -19,7 +19,7 @@ const { deleteFile } = store
 
 const children = computed(() => {
   let filtered = props.filter ? props.files.children?.filter(f => f.children || props.filter.test(f.name)) : props.files.children
-  return _sortBy(filtered, ['children', 'name'])
+  return _sortBy(filtered, [function(o) { return !o.children }, 'name'])
 })
 </script>
 

--- a/packages/electron/components/global/TntElectron/DirectoryTreeFileName.vue
+++ b/packages/electron/components/global/TntElectron/DirectoryTreeFileName.vue
@@ -1,0 +1,35 @@
+<script setup>
+import { useDirectoryStore } from '@/stores/directory'
+
+const props = defineProps([
+  'path',
+  'name'
+])
+
+const value = ref(props.name)
+
+const editing = ref(false)
+
+// Store
+const store = useDirectoryStore()
+// Store: Actions
+const { renameFile } = store
+
+function rename() {
+  const regex = new RegExp(`${props.name}$`)
+  const newPath = props.path.replace(regex, value.value)
+  renameFile(props.path, newPath)
+  editing.value = false
+}
+</script>
+
+<template lang="pug">
+TntForm.inline(v-if="editing" @submit="rename()" @keydown.esc="editing = false")
+  TntInput.inline-block(v-model="value")
+  TntSubmit.btn-none.ml-2
+    Icon(name="fa6-solid:floppy-disk")
+template(v-else)
+  span {{ name }}
+  TntButton.btn-none(@click.prevent="editing = true")
+    Icon.ml-2(name="fa6-solid:pencil")
+</template>

--- a/packages/electron/components/global/TntElectron/DirectoryTreeFileName.vue
+++ b/packages/electron/components/global/TntElectron/DirectoryTreeFileName.vue
@@ -9,7 +9,7 @@ const props = defineProps([
   'link',
 ])
 
-const initialName = ref(name.value)
+const newName = ref(name.value)
 
 const editing = ref(false)
 
@@ -21,23 +21,23 @@ const { root } = storeToRefs(store)
 const { renameFile } = store
 
 function rename() {
-  const regex = new RegExp(`${initialName.value}$`)
-  const newPath = path.value.replace(regex, name.value)
+  const regex = new RegExp(`${name.value}$`)
+  const newPath = path.value.replace(regex, newName.value)
   renameFile(path.value, newPath)
   path.value = newPath
-  initialName.value = name.value
+  name.value = newName.value
   editing.value = false
 }
 
 function cancel() {
-  name.value = initialName.value
+  newName.value = name.value
   editing.value = false
 }
 </script>
 
 <template lang="pug">
 TntForm.inline(v-if="editing" @submit="rename()" @keydown.esc="cancel()")
-  TntInput.inline-block(v-model="name")
+  TntInput.inline-block(v-model="newName")
   TntSubmit.btn-none.ml-2
     Icon(name="fa6-solid:floppy-disk")
 template(v-else)

--- a/packages/electron/components/global/TntElectron/DirectoryTreeFileName.vue
+++ b/packages/electron/components/global/TntElectron/DirectoryTreeFileName.vue
@@ -1,9 +1,11 @@
 <script setup>
 import { useDirectoryStore } from '@/stores/directory'
+import { storeToRefs } from 'pinia'
 
 const props = defineProps([
   'path',
-  'name'
+  'name',
+  'link',
 ])
 
 const value = ref(props.name)
@@ -12,6 +14,8 @@ const editing = ref(false)
 
 // Store
 const store = useDirectoryStore()
+// Getters
+const { root } = storeToRefs(store)
 // Store: Actions
 const { renameFile } = store
 
@@ -29,7 +33,12 @@ TntForm.inline(v-if="editing" @submit="rename()" @keydown.esc="editing = false")
   TntSubmit.btn-none.ml-2
     Icon(name="fa6-solid:floppy-disk")
 template(v-else)
-  span {{ name }}
-  TntButton.btn-none(@click.prevent="editing = true")
-    Icon.ml-2(name="fa6-solid:pencil")
+  NuxtLink(v-if="link" :to="path.replace(`${root}/`, '')")
+    span {{ name }}
+    TntButton.btn-none(@click.prevent="editing = true")
+      Icon.ml-2(name="fa6-solid:pencil")
+  template(v-else)
+    span {{ name }}
+    TntButton.btn-none(@click.prevent="editing = true")
+      Icon.ml-2(name="fa6-solid:pencil")
 </template>

--- a/packages/electron/components/global/TntElectron/DirectoryTreeFileName.vue
+++ b/packages/electron/components/global/TntElectron/DirectoryTreeFileName.vue
@@ -20,25 +20,34 @@ const { root } = storeToRefs(store)
 const { renameFile } = store
 
 function rename() {
+  // TODO: This breaks for subsequent edits - name prop is not reactive
+  // TODO: Nested items also cannot be updated if the directory name has also already been changed
+  //       Again, the props aren't reactive!
   const regex = new RegExp(`${props.name}$`)
   const newPath = props.path.replace(regex, value.value)
   renameFile(props.path, newPath)
   editing.value = false
 }
+
+function cancel() {
+  editing.value = false
+  // TODO: This breaks if already edited - name prop is not reactive
+  value.value = props.name
+}
 </script>
 
 <template lang="pug">
-TntForm.inline(v-if="editing" @submit="rename()" @keydown.esc="editing = false")
+TntForm.inline(v-if="editing" @submit="rename()" @keydown.esc="cancel()")
   TntInput.inline-block(v-model="value")
   TntSubmit.btn-none.ml-2
     Icon(name="fa6-solid:floppy-disk")
 template(v-else)
   NuxtLink(v-if="link" :to="path.replace(`${root}/`, '')")
-    span {{ name }}
+    span {{ value }}
     TntButton.btn-none(@click.prevent="editing = true")
       Icon.ml-2(name="fa6-solid:pencil")
   template(v-else)
-    span {{ name }}
+    span {{ value }}
     TntButton.btn-none(@click.prevent="editing = true")
       Icon.ml-2(name="fa6-solid:pencil")
 </template>

--- a/packages/electron/components/global/TntElectron/DirectoryTreeFileName.vue
+++ b/packages/electron/components/global/TntElectron/DirectoryTreeFileName.vue
@@ -2,13 +2,14 @@
 import { useDirectoryStore } from '@/stores/directory'
 import { storeToRefs } from 'pinia'
 
+const path = defineModel('path')
+const name = defineModel('name')
+
 const props = defineProps([
-  'path',
-  'name',
   'link',
 ])
 
-const value = ref(props.name)
+const initialName = ref(name.value)
 
 const editing = ref(false)
 
@@ -20,34 +21,32 @@ const { root } = storeToRefs(store)
 const { renameFile } = store
 
 function rename() {
-  // TODO: This breaks for subsequent edits - name prop is not reactive
-  // TODO: Nested items also cannot be updated if the directory name has also already been changed
-  //       Again, the props aren't reactive!
-  const regex = new RegExp(`${props.name}$`)
-  const newPath = props.path.replace(regex, value.value)
-  renameFile(props.path, newPath)
+  const regex = new RegExp(`${initialName.value}$`)
+  const newPath = path.value.replace(regex, name.value)
+  renameFile(path.value, newPath)
+  path.value = newPath
+  initialName.value = name.value
   editing.value = false
 }
 
 function cancel() {
+  name.value = initialName.value
   editing.value = false
-  // TODO: This breaks if already edited - name prop is not reactive
-  value.value = props.name
 }
 </script>
 
 <template lang="pug">
 TntForm.inline(v-if="editing" @submit="rename()" @keydown.esc="cancel()")
-  TntInput.inline-block(v-model="value")
+  TntInput.inline-block(v-model="name")
   TntSubmit.btn-none.ml-2
     Icon(name="fa6-solid:floppy-disk")
 template(v-else)
   NuxtLink(v-if="link" :to="path.replace(`${root}/`, '')")
-    span {{ value }}
+    span {{ name }}
     TntButton.btn-none(@click.prevent="editing = true")
       Icon.ml-2(name="fa6-solid:pencil")
   template(v-else)
-    span {{ value }}
+    span {{ name }}
     TntButton.btn-none(@click.prevent="editing = true")
       Icon.ml-2(name="fa6-solid:pencil")
 </template>

--- a/packages/electron/electron/api.ts
+++ b/packages/electron/electron/api.ts
@@ -77,14 +77,14 @@ export default function initIpc() {
     })
   })
 
-  ipcMain.handle('rename-file', (_, path:string, name:string) => {
+  ipcMain.handle('rename-file', (_, path:string, name:string):Promise<void> => {
     return new Promise((resolve, reject) => {
-      fs.rename(join(String(process.env.PORTABLE_EXECUTABLE_DIR || ""), path), name, (error) => {
+      fs.rename(join(String(process.env.PORTABLE_EXECUTABLE_DIR || ""), path), name, function (error) {
         if (error) {
           reject(error)
           return
         }
-        return
+        resolve()
       })
     })
   })

--- a/packages/electron/electron/api.ts
+++ b/packages/electron/electron/api.ts
@@ -118,4 +118,16 @@ export default function initIpc() {
       })
     })
   })
+
+  ipcMain.on('create-folder', (_, path):Promise<void> => {
+    return new Promise((resolve, reject) => {
+      fs.mkdir(join(String(process.env.PORTABLE_EXECUTABLE_DIR || ""), path), { recursive: true }, (error) => {
+        if (error) {
+          reject(error)
+          return
+        }
+        resolve()
+      })
+    })
+  })
 }

--- a/packages/electron/electron/api.ts
+++ b/packages/electron/electron/api.ts
@@ -105,7 +105,7 @@ export default function initIpc() {
     })
   })
 
-  ipcMain.on('update-file', (_, path, content) => {
+  ipcMain.on('update-file', (_, path, content:string = "") => {
     return new Promise((resolve, reject) => {
       fs.readFile(join(String(process.env.PORTABLE_EXECUTABLE_DIR || ""), path), "utf8", (error, file) => {
         fs.writeFile(join(String(process.env.PORTABLE_EXECUTABLE_DIR || ""), path), content, (error) => {

--- a/packages/electron/electron/api.ts
+++ b/packages/electron/electron/api.ts
@@ -1,4 +1,4 @@
-import { ipcMain } from 'electron'
+import { ipcMain, shell } from 'electron'
 import fs from 'fs'
 import { join } from 'path'
 
@@ -74,6 +74,10 @@ export default function initIpc() {
         resolve({} as DirectoryTree)
       }
     })
+  })
+
+  ipcMain.handle('delete-file', (_, path:string):Promise<void> => {
+    return shell.trashItem(join(String(process.env.PORTABLE_EXECUTABLE_DIR || ""), path))
   })
 
   ipcMain.handle('load-file', (_, path) => {

--- a/packages/electron/electron/api.ts
+++ b/packages/electron/electron/api.ts
@@ -3,7 +3,6 @@ import fs from 'fs'
 import { join } from 'path'
 
 import dirTree, { DirectoryTree } from 'directory-tree'
-import { stringify } from 'querystring'
 
 export default function initIpc() {
   ipcMain.handle('load-config', () => {

--- a/packages/electron/electron/api.ts
+++ b/packages/electron/electron/api.ts
@@ -3,6 +3,7 @@ import fs from 'fs'
 import { join } from 'path'
 
 import dirTree, { DirectoryTree } from 'directory-tree'
+import { stringify } from 'querystring'
 
 export default function initIpc() {
   ipcMain.handle('load-config', () => {
@@ -73,6 +74,18 @@ export default function initIpc() {
       } else {
         resolve({} as DirectoryTree)
       }
+    })
+  })
+
+  ipcMain.handle('rename-file', (_, path:string, name:string) => {
+    return new Promise((resolve, reject) => {
+      fs.rename(join(String(process.env.PORTABLE_EXECUTABLE_DIR || ""), path), name, (error) => {
+        if (error) {
+          reject(error)
+          return
+        }
+        return
+      })
     })
   })
 

--- a/packages/electron/stores/directory.ts
+++ b/packages/electron/stores/directory.ts
@@ -24,7 +24,7 @@ export const useDirectoryStore = defineStore('directory', () => {
   }
   
   async function createFile(file:string) {
-    useTntApi().updateFile(file)
+    await useTntApi().updateFile(file)
     toasts.value.push({
       uid: _uniqueId('tnt-toast-create-file-'),
       duration: 5000,

--- a/packages/electron/stores/directory.ts
+++ b/packages/electron/stores/directory.ts
@@ -39,6 +39,23 @@ export const useDirectoryStore = defineStore('directory', () => {
     })
     fetchDirectory(root.value)
   }
+  
+  async function createFolder(folder:string) {
+    await useTntApi().createFolder(folder)
+    toasts.value.push({
+      uid: _uniqueId('tnt-toast-create-folder-'),
+      duration: 5000,
+      color: 'success',
+      body: [
+        { is: 'strong', props: { class: 'text-lg' }, content: 'Added' },
+        { is: 'p', content: `A folder has been created at ${folder}.` },
+      ],
+      actions: [
+        { is: 'Icon', props: { class: 'text-2xl', name: 'fa6-solid:folder' } }
+      ]
+    })
+    fetchDirectory(root.value)
+  }
 
   async function renameFile(file:string, name:string) {
     await useTntApi().renameFile(file, name)
@@ -82,6 +99,7 @@ export const useDirectoryStore = defineStore('directory', () => {
     // Actions
     fetchDirectory,
     createFile,
+    createFolder,
     renameFile,
     deleteFile,
   }

--- a/packages/electron/stores/directory.ts
+++ b/packages/electron/stores/directory.ts
@@ -30,7 +30,7 @@ export const useDirectoryStore = defineStore('directory', () => {
       color: 'danger',
       body: [
         { is: 'strong', props: { class: 'text-lg' }, content: 'Deleted' },
-        { is: 'p', content: 'This file has been sent to your trash folder.' },
+        { is: 'p', content: `${file} has been sent to your trash folder.` },
       ],
       actions: [
         { is: 'Icon', props: { class: 'text-2xl', name: 'fa6-solid:trash-can' } }

--- a/packages/electron/stores/directory.ts
+++ b/packages/electron/stores/directory.ts
@@ -22,6 +22,23 @@ export const useDirectoryStore = defineStore('directory', () => {
     const dir = await useTntApi().listFiles(path, opts)
     if (dir) tree.value = dir
   }
+  
+  async function createFile(file:string) {
+    useTntApi().updateFile(file)
+    toasts.value.push({
+      uid: _uniqueId('tnt-toast-create-file-'),
+      duration: 5000,
+      color: 'success',
+      body: [
+        { is: 'strong', props: { class: 'text-lg' }, content: 'Added' },
+        { is: 'p', content: `A file has been created at ${file}.` },
+      ],
+      actions: [
+        { is: 'Icon', props: { class: 'text-2xl', name: 'fa6-solid:file' } }
+      ]
+    })
+    fetchDirectory(root.value)
+  }
 
   async function renameFile(file:string, name:string) {
     await useTntApi().renameFile(file, name)
@@ -64,6 +81,7 @@ export const useDirectoryStore = defineStore('directory', () => {
     root,
     // Actions
     fetchDirectory,
+    createFile,
     renameFile,
     deleteFile,
   }

--- a/packages/electron/stores/directory.ts
+++ b/packages/electron/stores/directory.ts
@@ -8,6 +8,8 @@ import _uniqueId from 'lodash/uniqueId'
 
 import { useToasts } from '@thombruce/tnt/composables/states'
 
+const toasts = useToasts()
+
 export const useDirectoryStore = defineStore('directory', () => {
   // State
   const tree = ref({} as DirectoryTree)
@@ -21,8 +23,24 @@ export const useDirectoryStore = defineStore('directory', () => {
     if (dir) tree.value = dir
   }
 
+  async function renameFile(file:string, name:string) {
+    await useTntApi().renameFile(file, name)
+    toasts.value.push({
+      uid: _uniqueId('tnt-toast-rename-file-'),
+      duration: 5000,
+      color: 'primary',
+      body: [
+        { is: 'strong', props: { class: 'text-lg' }, content: 'Renamed' },
+        { is: 'p', content: `${file} has been renamed to ${name}.` },
+      ],
+      actions: [
+        { is: 'Icon', props: { class: 'text-2xl', name: 'fa6-solid:info' } }
+      ]
+    })
+    fetchDirectory(root.value)
+  }
+
   async function deleteFile(file:string) {
-    const toasts = useToasts()
     await useTntApi().deleteFile(file)
     toasts.value.push({
       uid: _uniqueId('tnt-toast-delete-file-'),
@@ -46,6 +64,7 @@ export const useDirectoryStore = defineStore('directory', () => {
     root,
     // Actions
     fetchDirectory,
+    renameFile,
     deleteFile,
   }
 })

--- a/packages/electron/stores/directory.ts
+++ b/packages/electron/stores/directory.ts
@@ -1,0 +1,51 @@
+import { ref, computed } from 'vue'
+import { defineStore } from 'pinia'
+
+import { useTntApi } from '@thombruce/tnt/composables/tntApi'
+import { DirectoryTree } from 'directory-tree'
+
+import _uniqueId from 'lodash/uniqueId'
+
+import { useToasts } from '@thombruce/tnt/composables/states'
+
+export const useDirectoryStore = defineStore('directory', () => {
+  // State
+  const tree = ref({} as DirectoryTree)
+
+  // Getters
+  const root = computed(() => tree.value.path)
+
+  // Actions
+  async function fetchDirectory(path:string, opts?:object) {
+    const dir = await useTntApi().listFiles(path, opts)
+    if (dir) tree.value = dir
+  }
+
+  async function deleteFile(file:string) {
+    const toasts = useToasts()
+    await useTntApi().deleteFile(file)
+    toasts.value.push({
+      uid: _uniqueId('tnt-toast-delete-file-'),
+      duration: 5000,
+      color: 'danger',
+      body: [
+        { is: 'strong', props: { class: 'text-lg' }, content: 'Deleted' },
+        { is: 'p', content: 'This file has been sent to your trash folder.' },
+      ],
+      actions: [
+        { is: 'Icon', props: { class: 'text-2xl', name: 'fa6-solid:trash-can' } }
+      ]
+    })
+    fetchDirectory(root.value)
+  }
+
+  return {
+    // State
+    tree,
+    // Getters
+    root,
+    // Actions
+    fetchDirectory,
+    deleteFile,
+  }
+})

